src/parser/SparqlParser.cpp

```diff
@@ -417,119 +417,91 @@ std::string_view SparqlParser::readTriplePart(const std::string& s,
 
 // _____________________________________________________________________________
 void SparqlParser::parseSolutionModifiers(ParsedQuery* query) {
-  while (!lexer_.empty() && !lexer_.accept("}")) {
-    if (lexer_.peek(SparqlToken::Type::ORDER_BY)) {
-      auto order_keys =
-          parseWithAntlr(sparqlParserHelpers::parseOrderClause, *query);
-
-      auto processVariableOrderKey = [&query](VariableOrderKey orderKey) {
-        // Check whether grouping is done. The variable being ordered by
-        // must then be either grouped or the result of an alias in the select.
-        const vector<Variable>& groupByVariables = query->_groupByVariables;
-        if (!groupByVariables.empty() &&
-            !ad_utility::contains_if(groupByVariables,
-                                     [&orderKey](const Variable& var) {
-                                       return orderKey.variable_ == var.name();
-                                     }) &&
-            !ad_utility::contains_if(query->selectClause().getAliases(),
-                                     [&orderKey](const Alias& alias) {
-                                       return alias._outVarName ==
-                                              orderKey.variable_;
-                                     })) {
-          throw ParseException(
-              "Variable " + orderKey.variable_ +
-              " was used in an ORDER BY "
-              "clause, but is neither grouped, nor created as an alias in the "
-              "SELECT clause.");
-        }
-
-        query->_orderBy.push_back(std::move(orderKey));
-      };
+  auto modifiers =
+      parseWithAntlr(sparqlParserHelpers::parseSolutionModifier, *query);
 
-      // QLever currently only supports ordering by variables. To allow
-      // all `orderConditions`, the corresponding expression is bound to a new
-      // internal variable. Ordering is then done by this variable.
-      auto processExpressionOrderKey = [&query,
-                                        this](ExpressionOrderKey orderKey) {
-        if (!query->_groupByVariables.empty())
-          // TODO<qup42> Implement this by adding a hidden alias in the
-          //  SELECT clause.
-          throw ParseException(
-              "Ordering by an expression while grouping is not supported by "
-              "QLever. (The expression is \"" +
-              orderKey.expression_.getDescriptor() +
-              "\"). Please assign this expression to a "
-              "new variable in the SELECT clause and then order by this "
-              "variable.");
-        auto additionalVariable =
-            addInternalBind(query, std::move(orderKey.expression_));
-        query->_orderBy.emplace_back(additionalVariable.name(),
-                                     orderKey.isDescending_);
+  // Process groupClause
+  auto processVariable = [&query](const Variable& groupKey) {
```

Comment:
I need some more time to think about how to do this properly. If implemented directly, we would need to pass the internal variable state through `addSolutionModifiers` to `addInternalBind`.

Reply:
I see what you mean.
The obvious solution is, that `addInternalBind` should also be a member function of `ParsedQuery`. It only uses the `numInternalVariables` member of the SparqlParser which could also become part of the `ParsedQuery`. But this is maybe then a (small and therefore nicely reviewable) PR on its own, so I would be fine with leaving this as is for now.

Reply:
Yes but would this work? This would change the `numInternalVariables` from being global to being local to a (sub-)query. I'll look into whether that works. If it does not work without further changes I'll leave it for another PR.